src/org/freenetproject/freemail/ui/web/WebPage.java

```diff
@@ -177,6 +183,41 @@ protected void addWoTNotLoadedMessage(HTMLNode parent) {
 				new HTMLNode[] {HTMLNode.link("/plugins")});
 	}
 
+	void addChild(HTMLNode parent, String templateName, Map<String, String> model) throws IOException {
+		try (InputStream stream =
+						getClass().getResourceAsStream("/templates/" + templateName + ".html")) {
+			ByteArrayOutputStream content = new ByteArrayOutputStream();
+
+			int len;
+			byte[] contentBytes = new byte[1024];
+			while ((len = stream.read(contentBytes)) != -1)
+				content.write(contentBytes, 0, len);
+
+			String template = content.toString(StandardCharsets.UTF_8.name());
+
+			for (Map.Entry<String, String> entry : model.entrySet())
+				template = template.replaceAll("\\$\\{" + entry.getKey() + "}", entry.getValue());
```

Comment:
While I appreciate your effort here to create something that is more easy to write than hand-coded HTML using `HTMLNode`, this is not it.
* The keys are not escaped correctly.
* The values are not escaped correctly.
* Creating your own template engine is a rabbit hole.
* The template is parsed for every request.
* It will not help to improve the rest of the web interface, neither for Freemail, nor for the rest of Fred.
As horrible as it is (and it is very much so), please use `HTMLNode` for this.

Reply:
Thanks a lot for your feedback. 
I understand it is a big decision to make, but I think we all agree that Fred and the plugins are quite hard to maintain. It appears the issue grows bigger as newer approaches are adopted by other projects, which may hurt the Freenet tech and community in the long run, so at some point a decision on modernizing the project one way or another would have to be made.  

Introducing a template-based approach feels like a good way forward to us. I got the idea to create a simple template engine from KeepAlive plugin and I guess many other coders would be happy with such approach, since template engine would help avoid vulnerabilities linked to using external libraries, while allowing to make changes to the project more efficiently. 

However, perhaps you have other options or suggestions in mind? Please let me know, we would be happy to assist you with them.

Reply:
I don’t have a problem with template engines, they would totally solve quite a number of problems we have in Fred. :) However, rolling your own broken template engine is not the solution.

I also don’t think that external dependencies are “evil” and need to be avoided at all cost… however, picking one for Fred and using that everywhere will incur the wrath of everybody who has ever had a slightly negative experience with whatever we picked, and we would have to maintain the included version, patching/updating it without ever breaking anything else… yeah so I think that approach is also doomed.

I think my favoured approach would be that every plugin/app has to take care of its own HTML generation; I can totally live with a custom implementation in Freemail, provided it is not trivial to break and doesn’t needlessly parse the same three patterns over and over again. :)